### PR TITLE
(PDB-4357) FIPS support via bc-fips jar

### PR DESCRIPTION
**Pull request by robdaemon**
Move the bouncycastle jar to the dev profile. This will allow us
to specify the bouncycastle jars on the classpath at runtime like
we do for logging configuration. At installation time, we can install
the FIPS-compliant version of the jars, or the standard versions.

Remove any and all SSLv3 references / tests. FIPS does not allow SSLv3,
and the JVM explicitly disables SSLv3. OWASP guidelines specifically
state:

    SSL/TLS is a collection of protocols. Weaknesses have been identified with earlier SSL protocols, including SSLv2 and SSLv3, hence SSL versions 1, 2, and 3 should not longer be used. The best practice for transport layer protection is to only provide support for the TLS protocols - TLS 1.0, TLS 1.1 and TLS 1.2

Removing the test.sh script in favor of calling lein directly.

Adding CI tests for JDK 8 with and without FIPS.

### CODE REVIEW

**Comment by robdaemon** [comment]
depends on puppetlabs/jvm-ssl-utils#87

File: .travis.yml [hunk 1]
```diff
@@ -1,11 +1,21 @@
 language: clojure
-lein: 2.8.1
-jdk:
-  - oraclejdk8
-script: ./ext/travisci/test.sh
-notifications:
-  email: false
```
Comment (justinstoller) [inline .travis.yml]: Are email notifications no longer enabled by default anymore?

File: project.clj [hunk 2]
```diff
@@ -38,9 +38,17 @@
                                   [puppetlabs/trapperkeeper nil :classifier "test"]
                                   [puppetlabs/trapperkeeper-webserver-jetty9]
                                   [puppetlabs/trapperkeeper-webserver-jetty9 nil :classifier "test"]
-                                  [puppetlabs/ring-middleware]]
-                   ;; TK-143, enable SSLv3 for unit tests that exercise SSLv3
-                   :jvm-opts ["-Djava.security.properties=./dev-resources/java.security"]}
+                                  [puppetlabs/ring-middleware]
+                                  [org.bouncycastle/bcpkix-jdk15on]]}
+             :fips {:dependencies [[cheshire]
+                                   [puppetlabs/kitchensink nil :classifier "test"]
```
Comment (justinstoller) [inline project.clj]: We no longer need to include the `nil` in profile dependencies. The bug that required that was fixed in leiningen 2.8.0. When we bump the travis lein version can/should we bump the `min-lein-version` in this file and remove the `nil`s?

File: test/puppetlabs/http/client/async_unbuffered_test.clj [hunk 3]
```diff
@@ -144,7 +145,7 @@
        (let [response @(common/get client (str "http://localhost:" 12345 "/bad") opts)
              {:keys [error]} response]
          (is error)
-         (is (instance? ConnectException error)))))))
+         (is (instance? ConnectionClosedException error)))))))
```
Comment (justinstoller) [inline test/puppetlabs/http/client/async_unbuffered_test.clj]: This seems like a weird change, do you know what instigated it?

File: test/puppetlabs/http/client/sync_ssl_test.clj [hunk 4]
```diff
@@ -161,27 +161,6 @@
           (is (= 200 (:status response)))
           (is (= "Hello, World!" (slurp (:body response))))))))
 
-  (testing "should be able to connect to a server with non-default protocol if configured"
```
Comment (justinstoller) [inline test/puppetlabs/http/client/sync_ssl_test.clj]: Using SSLv3 in these test seems like an incidental part of the implementation in three of them and the fourth is validating that we _don't_ negotiate down to SSLv3 even if a server allows it. None of them seem like things that need to be removed for enabling fips support, though I could see some, if not all of them, being low value compared to maintenance costs. Am I missing something about that?

If their relative value was your rationale I didn't catch that from the commit message/can we add it?

**Comment by justinstoller** [comment]
> Remove any and all SSLv3 references / tests. FIPS does not allow SSLv3,
and the JVM explicitly disables SSLv3. OWASP guidelines specifically
state:

I understand that fips does not allow SSLv3, and users _shouldn't_ be using SSLv3 internal to their networks even, though I don't know if we can remove the _option_ to use SSLv3 from non-fips users in the Puppet Platform 6.x series.

Given that, I'm not clear if our SSLv3 tests are doing much more than testing the upstream apache library honors its own configuration (assuming other tests validate that we are generally integrating with the library's configuration objects correctly).

**Comment by justinstoller** [comment]
I've integrated this, the requisite jvm-ssl-utils changes, and bc-fips libraries into a dev environment of puppetserver and validated all of our clojure based unit and integration tests pass. I'm confident in the changes here from a functional standpoint though I have a few housekeeping nits that I'd like addressed (and addressing some of them may simply require further discussion).

File: project.clj [hunk 5]
```diff
@@ -53,5 +61,8 @@
   :lein-release {:scm :git
                  :deploy-via :lein-deploy}
 
-  :plugins [[lein-parent "0.3.1"]
-            [puppetlabs/i18n "0.8.0"]])
+  :plugins [[lein-parent "0.3.5"]
```
Comment (justinstoller) [inline project.clj]: One other "give a mouse a cookie" kinda request. I'm working with achin to release lein-parent 0.3.7 which will remove some warnings that are output when using it with newer lein versions. It'd be nice to wait until Monday and then bump this to 0.3.7 (see https://github.com/achin/lein-parent/pull/20) though that can totally be done in a separate PR if needed.

File: .travis.yml (hunk 1 above)
Comment (robdaemon) [reply .travis.yml]: Oh, I'm not sure. I was just matching another project.

File: project.clj (hunk 2 above)
Comment (robdaemon) [reply project.clj]: Yeah, no worries.

File: test/puppetlabs/http/client/async_unbuffered_test.clj (hunk 3 above)
Comment (robdaemon) [reply test/puppetlabs/http/client/async_unbuffered_test.clj]: A change in commons-httpclient. The exception thrown is now `ConnectionClosedException` - this test did not work properly without this change. I'm assuming we updated the commons-httpclient version in clj-parent at some point but did not rerun the tests in this project?

File: test/puppetlabs/http/client/sync_ssl_test.clj (hunk 4 above)
Comment (robdaemon) [reply test/puppetlabs/http/client/sync_ssl_test.clj]: The behavior of these SSLv3 tests is radically different in a bc-fips env. The errors are different and everything.

Considering SSLv3 is disabled at the JRE level, I don't see any value in attempting to override those settings, so this felt almost "vestigial" for lack of a better term.

File: project.clj (hunk 5 above)
Comment (robdaemon) [reply project.clj]: Yeah, we can do that

File: .travis.yml (hunk 1 above)
Comment (justinstoller) [reply .travis.yml]: I'd rather not have the email notifications on: https://docs.travis-ci.com/user/notifications/#default-notification-settings

File: project.clj [hunk 6]
```diff
@@ -20,7 +20,7 @@
                  [commons-io]
                  [io.dropwizard.metrics/metrics-core]
 
-                 [puppetlabs/ssl-utils]
+                 [puppetlabs/ssl-utils "2.0.1-SNAPSHOT"]
```
Comment (justinstoller) [inline project.clj]: We've release 3.0.1 with the needed changes in it. I suppose we should do the clj-parent dance so we can just remove the explicit dependency here.

**Comment by justinstoller** [comment]
Hey @robdaemon we've got a 4.0.0 of clj-parent out that has the jvm-ssl-utils bump in it! If you can work that in here then we can do a 4.1 and announce FIPS support in clj-parent!

**Comment by robdaemon** [comment]
@justinstoller done!

**Comment by robdaemon** [comment]
ugh yay new test failures I hadn't seen before, gimme some time today to clean this up more

**Comment by robdaemon** [comment]
Okay I took JDK 11 testing out of this. Things seem to behave very differently under JDK 11, so that needs to be fixed as a separate story.
